sympy/functions/elementary/integers.py

```diff
@@ -132,11 +132,21 @@ def _eval_number(cls, arg):
             return arg.approximation_interval(Integer)[0]
 
     def _eval_as_leading_term(self, x, logx=None, cdir=0):
-        arg = self.args[0]
+        arg = self.args[0].as_leading_term(x, logx=logx, cdir=cdir)
         arg0 = arg.subs(x, 0)
+        if arg0 is S.NaN:
+            arg0 = arg.limit(x, 0, dir='-' if cdir.is_negative else '+')
+        r = self._eval_nseries(x, n=1, logx=None, cdir=cdir)
```

Comment:
I would have expected something like this (without the comments):
```
    r = self.subs(x, 0)  # Can this be nan? Probably not in the current examples. 
    if arg0 == r:  # arg0 is an integer
        ndir = arg.dir(x, cdir)  # new direction depending on arg
        return r - 1 if ndir.is_negative else r
```
`dir` is cheaper than `leadterm`. Besides, `leadterm` is usually independent of `cdir` so the coefficient it returns may have wrong sign. (Consider `arg = arg0 + x`: `(arg - arg0).leadterm(x, cdir=cdir)` is `(1, 1)`.)

Reply:
The problem is that `x.dir(x, cdir)` and `(-x).dir(x, cdir)` are both `0`, when `cdir=0`

Reply:
We should try to make `cdir` nonzero in a case like this where it would matter. I think that is how it is done in `Limit.doit`. Anyway, we could probably set `cdir = 1` if it is zero.

Reply:
We could also raise if `arg0 == r` and `cdir == 0` as then there is not twosided limit.

Reply:
In that case, should `ceiling(x).series(x)` raise? And what error should it raise, `ValueError`?

Reply:
It seems that a slightly more careful investigation is needed if `cdir == 0`, We could have something like `limit(floor(-x**2, x, 0, dir='+-')`. I think that both `arg.dir(x, 1)` and `arg.dir(x, -1)` should be computed and an error raised only if they disagree. Otherwise the common value can be used in the sequel.